arbnode/message_pruner.go

```diff
@@ -0,0 +1,187 @@
+
+	"github.com/offchainlabs/nitro/staker"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+
+	flag "github.com/spf13/pflag"
+)
+
+type MessagePruner struct {
+	stopwaiter.StopWaiter
+	transactionStreamer *TransactionStreamer
+	inboxTracker        *InboxTracker
+	staker              *staker.Staker
+	config              MessagePrunerConfigFetcher
+}
+
+type MessagePrunerConfig struct {
+	Enable               bool          `koanf:"enable"`
+	MessagePruneInterval time.Duration `koanf:"prune-interval" reload:"hot"`
+}
+
+type MessagePrunerConfigFetcher func() *MessagePrunerConfig
+
+var DefaultMessagePrunerConfig = MessagePrunerConfig{
+	Enable:               false,
+	MessagePruneInterval: time.Minute,
+}
+
+func MessagePrunerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultMessagePrunerConfig.Enable, "enable message pruning")
+	f.Duration(prefix+".prune-interval", DefaultMessagePrunerConfig.MessagePruneInterval, "interval for running message pruner")
+}
+
+func NewMessagePruner(transactionStreamer *TransactionStreamer, inboxTracker *InboxTracker, staker *staker.Staker, config MessagePrunerConfigFetcher) *MessagePruner {
+	return &MessagePruner{
+		transactionStreamer: transactionStreamer,
+		inboxTracker:        inboxTracker,
+		staker:              staker,
+		config:              config,
+	}
+}
+
+func (m *MessagePruner) Start(ctxIn context.Context) {
+	m.StopWaiter.Start(ctxIn, m)
+	m.CallIteratively(m.prune)
+}
+
+func (m *MessagePruner) prune(ctx context.Context) time.Duration {
+	latestConfirmedNode, err := m.staker.Rollup().LatestConfirmed(
+		&bind.CallOpts{
+			Context:     ctx,
+			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
+		})
+	if err != nil {
+		log.Error("error getting latest confirmed node: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	nodeInfo, err := m.staker.Rollup().LookupNode(ctx, latestConfirmedNode)
+	if err != nil {
+		log.Error("error getting latest confirmed node info: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	endBatchCount := nodeInfo.Assertion.AfterState.GlobalState.Batch
+	if endBatchCount == 0 {
+		return m.config().MessagePruneInterval
+	}
+	endBatchMetadata, err := m.inboxTracker.GetBatchMetadata(endBatchCount - 1)
+	if err != nil {
+		log.Error("error getting last batch metadata: %w", err)
+		return m.config().MessagePruneInterval
+	}
+	deleteOldMessageFromDB(endBatchCount, endBatchMetadata, m.inboxTracker.db, m.transactionStreamer.db)
+	return m.config().MessagePruneInterval
+}
+
+func deleteOldMessageFromDB(endBatchCount uint64, endBatchMetadata BatchMetadata, inboxTrackerDb ethdb.Database, transactionStreamerDb ethdb.Database) {
+	startBatchCount := uint64(1)
+	hasKey, err := inboxTrackerDb.Has(lastPrunedSequencerBatchMetaKey)
+	if err != nil {
+		log.Warn("error checking last pruned batch metadata: %w", err)
+	} else if hasKey {
+		data, err := inboxTrackerDb.Get(lastPrunedSequencerBatchMetaKey)
```

Comment:
Instead of a separate database key, let's just use an iterator to get the first batch metadata key (this also applies to messages and delayed messages).

Reply:
done.